File: 104_mergeKLists.cpp
/*
104 合并k个排序链表

合并k个排序链表，并且返回合并后的排序链表。尝试分析和描述其复杂度。
样例
给出3个排序链表[2->4->null,null,-1->null]，返回 -1->2->4->null
*/

/**
 * Definition of ListNode
 * class ListNode {
 * public:
 *     int val;
 *     ListNode *next;
 *     ListNode(int val) {
 *         this->val = val;
 *         this->next = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     * @param lists: a list of ListNode
     * @return: The head of one sorted list.
     */
    ListNode *mergeKLists(vector<ListNode *> &lists) {
        // write your code here
        if(lists.empty())
        {
            return NULL;
        }//if
        
        int n = lists.size();
        ListNode *head = lists[0];
        for(int i=1; i<n; ++i)
        {
            head = merge(head,lists[i]);
        }//for
        
        return head;
        
    }
    
    ListNode *merge(ListNode *head1, ListNode *head2)
    {
        if(head1 == NULL)
        {
            return head2;
        }//if
        
        if(head2 == NULL)
        {
            return head1;
        }//if
        
        ListNode *p = head1 , *q = head2;
        ListNode *head = NULL, *r = head;
        
        while(p != NULL && q != NULL)
        {
            if(p->val <= q->val)
            {
                if(head == NULL)
                {
                    head = p;
                    r = head;
                }else {
                    r->next = p;
                    r = r->next;
                }//else
                p = p->next;
                r->next = NULL;
            }else{
                if(head == NULL)
                {
                    head = q;
                    r = head;
                }else{
                    r->next = q;
                    r = r->next;
                }//else
                q = q->next;
                r->next = NULL;
            }//else
        }//while
        
        if(p)
        {
            r->next = p;
        }//if
        
        if(q)
        {
            r->next = q;
        }//if
        
        return head;
    }
};

File: 104_mergeKLists.py
"""
Definition of ListNode
class ListNode(object):

    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""
class ListNode(object):
    def __init__(self,val,next=None):
        self.val = val
        self.next = next

def insert(head,val):
    node = ListNode(val)
    if head == None:
        return node

    p = head
    while p.next != None:
        p = p.next

    p.next = node
    return head

def display(head):
    p = head
    while p != None:
        print(p.val,end='\t')
        p = p.next
    print()


"""
Definition of ListNode
class ListNode(object):

    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""


# 合并两个链表
def merge(head1, head2):
    if head1 is None:
        return head2
    if head2 is None:
        return head1
    p = head1
    q = head2

    head = None
    r = head

    while p is not None and q is not None:
        if p.val <= q.val:
            if head is None:
                head = p
                r = head
            else:
                r.next = p
                r = r.next

            p = p.next
            r.next = None
        else:
            if head is None:
                head = q
                r = head
            else:
                r.next = q
                r = r.next

            q = q.next
            r.next = None

    if p is not None:
        r.next = p
    if q is not None:
        r.next = q
    return head


class Solution:
    """
    @param lists: a list of ListNode
    @return: The head of one sorted list.
    """
    def mergeKLists(self, lists):
        # write your code here
        n = len(lists)
        if n <= 0:
            return None
        head = lists[0]
        for i in range(1, n):
            head = merge(head, lists[i])
        return head


head1 = None

for i in range(1,6):
    head1 = insert(head1,i+1)

head2 = None

for i in range(1,6):
    head2 = insert(head2,i+2)

display(head1)
display(head2)

l = [head1,head2]

head = Solution().mergeKLists(l)

display(head)